### PR TITLE
reset uploading media to failed if the UploadService doesn't have them

### LINKED ISSUE #6455: Async media: inconsistent uploading items after crash
Originally reported by @daniloercoli 🙇 

### Expected behavior
When a crash or unexpected stop happens (turn then phone off, for example) while a Post with media is being uploaded, then coming back to the Post should show the interrupted items marked as failed and able to be retried.

### Actual behavior
These items don't show any specific mark and when tapped the "Stop uploading?" popup is displayed.

### Steps to reproduce the behavior
1. start a new draft
2. include some media items
3. while these are uploading, stop the app (go to settings-> apps -> WordPress -> force stop)
4. open the app again
5. open the draft
6. observe all media items do not show any "retry" mark. 
7.  tap on any failed item and observe the  "Stop uploading?" popup appears.

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #6455 

This PR does the following, under the conditions described in #6455:
Assuming you had some Post with media upload going on, or even a Post with failed media items in it, then you stop the app or turn the phone off/on.
- When you open the app again, there are no records of this having ever happened. It is only when you tap on the Post you were working on and open it in the Editor again now, that we're running through its content and check for any media item marked `UPLOADING`. These are all marked `FAILED` now, so they can be properly represented by the Editor with the failed / retry overlay. (*)
- Also, if we had any `UPLOADING` or `FAILED` items that for some reason already have a remote URL in their `src`  tags within the Post Content, these are unmarked as these are already valid URLs for the Post.
- There's still the case where you can see images or videos in the failed/retry state and tap on them to retry. This can happen when an image has been uploaded successfully, and so the `MediaStore` has the right information, but the `PostModel` has not yet been updated with the right information (that is, giving the `src` attribute the remote url value) when the crash / stop happens. In this case, the  user taps on the image, and the Retry callback will check whether the image already exist in the `MediaStore` and it has a remoteUrl, and just use it, recovering the image immediately). 
- **Note 1** (*): Given this is not a process we run for all posts given it could be a long running task, we rather do this only for each post as you try to open it in the editor. That's why there's no indication of previous media failed items on the Post list right after coming back from a crash/force stop. So, it is expected that you'll see your failed Post (the one you were working on before the crash) as a `Local draft` in the Posts list.
- **Note 2**: if you have several media items marked failed, and start uploading some of them (but not all), and go back, they will still be loading in the background, but the Post in the posts list will not show any progress, so giving more importance to failed media items that need action from the user (this was also the behavior before this PR).
If you start uploading them all (i.e. no failed items marked), we’ll be showing the progress as usual.

A better approach to this could be made, showing all failed Posts in the post list at once, recovering from any crash, once we[ move the `UploadStatus` handling to FluxC](https://github.com/wordpress-mobile/WordPress-Android/issues/4341#issuecomment-319413063). In the meantime, this PR tackles the problem described in #6455 which you cannot possibly recover from (blocking issue) without deleting all inconsistent media items and starting again.

To test:
1. start a new draft
2. include some media items
3. while these are uploading, stop the app (go to settings-> apps -> WordPress -> force stop)
4. open the app again
5. open the draft
6. observe failed media items show the failed/retry overlay
7.  tap on any failed item and observe the it either starts uploading again, or recovers its state gracefully if it's already found in the MediaStore
8. also tap on any non-failed item and observe the ImageSettings screen for such item is displayed, as expected.

cc @aforcier

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java [hunk 1]
```diff
@@ -378,6 +381,47 @@ public static PostModel updatePostWithCurrentlyFailedUploads(PostModel post) {
         return post;
     }
 
+    // this method aims at recovering the current state of media items if they're inconsistent within the PostModel.
+    public static void resetUploadingMediaToFailedIfNotInProgressOrQueued(Context context, PostModel post) {
```
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java]: Relying on Aztec status feels very out of place here - since this is only being called in `onCreate()` of the `EditPostActivity`, why not move this method into `AztecEditorFragment` (and expose any necessary `UploadService` methods so they can be called from there)?

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java]: :man_facepalming:  you're so much right, sorry - moved to `EditPostActivity` where it truly belongs, as within this method we also make use of `UploadService` and we can't reference that one from within `AztecEditorFragment`. In 096290b

**Comment by mzorz** [comment]
@daniloercoli I'd like you to review this one, if ok? :)

**Comment by daniloercoli** [comment]
Sure, assigned it to me.

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 2]
```diff
@@ -1452,4 +1467,89 @@ public static String markMediaFailed(Context context, @NonNull String postConten
...
+        return hasMediaItemsMarkedWithTag(context, postContent, ATTR_STATUS_UPLOADING);
+    }
+
+    public static boolean hasMediaItemsMarkedFailed(Context context, @NonNull String postContent) {
+        return hasMediaItemsMarkedWithTag(context, postContent, ATTR_STATUS_FAILED);
+    }
+
+    private static boolean hasMediaItemsMarkedWithTag(Context context, @NonNull String postContent, String tag) {
+        // fill in Aztec with the post's content
+        AztecText content = new AztecText(context);
```
Comment (daniloercoli) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: Wonder why if it's easier to just get all the media items in the editor and check for the presence of attribute. Without using Predicates.

```AztecMediaSpan[] imageOrVideoSpans = content.getText().getSpans(0, content.getText().length(), AztecMediaSpan.class);
                    for (AztecMediaSpan currentClass: imageOrVideoSpans) {
                      // do the check here
                    }
```

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 3]
```diff
@@ -1452,4 +1467,89 @@ public static String markMediaFailed(Context context, @NonNull String postConten
...
+        resetMediaWithStatus(content, ATTR_STATUS_UPLOADING);
+
+        // re-set the post content
+        postContent = content.toHtml(false);
+        return postContent;
+    }
+
+    private static void resetMediaWithStatus(AztecText content, String status) {
+        // get all items with "uploading" class
+        AztecText.AttributePredicate uploadingPredicate = getPredicateWithClass(status);
```
Comment (daniloercoli) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: Same comment of above ^ - not sure which of these 2 implementations is better (and also faster).

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java (hunk 2 above)
Comment (mzorz) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: ok I tried and measured the amount of time spent in the for loop versus using `content.getAllElementAttributes()` with the predicate, for a Post containing more than 50 images, both approaches resulted in very similar times, approximately 0 to 1 millisecond spent most of the times, with a 3 ms outlier (but still a non-significant value):

```
08-08 11:36:02.040 15486-15486/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 1
08-08 11:36:02.074 15486-15486/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 1
08-08 11:36:02.214 15486-15486/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 0
08-08 11:39:36.475 19524-19524/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 3
08-08 11:39:36.795 19524-19524/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 1
```

This is using the change proposed, measuring time before and after the code block containing both the `getSpans` call and the `for` loop to iterate through them:

```
08-08 11:42:01.968 20960-20960/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 0
08-08 11:42:02.052 20960-20960/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 1
08-08 11:48:24.001 20960-20960/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 0
08-08 11:48:24.049 20960-20960/org.wordpress.android E/WordPress-PROFILING: TOTAL TIME: 0
```

I think given in both cases the time is very similar and, in any case, not significant, I'd prefer to go with the code that uses predicates as for having a separation, i.e. so we don't need to look within the spans contained by the `AztecText`  component but rather using the exposed API.

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 4]
```diff
@@ -1491,4 +1506,89 @@ public static String markMediaFailed(Context context, @NonNull String postConten
...
+    public static boolean hasMediaItemsMarkedFailed(Context context, @NonNull String postContent) {
+        return hasMediaItemsMarkedWithTag(context, postContent, ATTR_STATUS_FAILED);
+    }
+
+    private static boolean hasMediaItemsMarkedWithTag(Context context, @NonNull String postContent, String tag) {
+        // fill in Aztec with the post's content
+        AztecText content = new AztecText(context);
+        content.fromHtml(postContent);
+
+        // get all items with "uploading" class
```
Comment (nbradbury) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: This comment appears incorrect - we're getting all the items with the passed class, not the "uploading" class.

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 5]
```diff
@@ -1491,4 +1506,89 @@ public static String markMediaFailed(Context context, @NonNull String postConten
...
+        AztecText.AttributePredicate uploadingPredicate = getPredicateWithClass(tag);
+
+        List<AztecAttributes> attrs = content.getAllElementAttributes(uploadingPredicate);
```
Comment (nbradbury) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: Based on your reply to Danilo it sounds like performance isn't a problem here, so feel free to ignore this if I'm being picky....but it strikes me as inefficient to get all element attributes here. Is there a way to do a `findFirst` to match only the first one and return true if one is found?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 6]
```diff
@@ -404,6 +406,40 @@ public void run() {
         ActivityId.trackLastActivity(ActivityId.POST_EDITOR);
     }
 
+    // this method aims at recovering the current state of media items if they're inconsistent within the PostModel.
+    private void resetUploadingMediaToFailedIfNotInProgressOrQueued(Context context, PostModel post) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: It seems unnecessary to pass the context and the post since the context is this activity and the post is `mPost`.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 7]
```diff
@@ -404,6 +406,40 @@ public void run() {
...
+        boolean useAztec = AppPrefs.isAztecEditorEnabled();
+        if (useAztec && context != null && post != null) {
+
+            String oldContent = post.getContent();
+            if (!UploadService.hasPendingOrInProgressMediaUploadsForPost(post) &&
+                    ( AztecEditorFragment.hasMediaItemsMarkedUploading(context, oldContent)
+                            || AztecEditorFragment.hasMediaItemsMarkedFailed(context, oldContent))) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Why is it necessary to check `hasMediaItemsMarkedFailed`?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 8]
```diff
@@ -404,6 +406,40 @@ public void run() {
         ActivityId.trackLastActivity(ActivityId.POST_EDITOR);
     }
 
+    // this method aims at recovering the current state of media items if they're inconsistent within the PostModel.
+    private void resetUploadingMediaToFailedIfNotInProgressOrQueued(Context context, PostModel post) {
+        boolean useAztec = AppPrefs.isAztecEditorEnabled();
+        if (useAztec && context != null && post != null) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I think we can get rid of some `if` nesting by doing a `return` if `useAztec` is false or there aren't any pending/in-progress uploads.

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java (hunk 4 above)
Comment (mzorz) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: addressed in fb63fb7

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java (hunk 5 above)
Comment (mzorz) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: some optimization could be done there by adding a new method in AztecText, but yeah I don't think it's necessary

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 6 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: oops! good catch - this was left there because it used not to belong to the Activity 😁 
addressed in 693b206

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 7 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: This is one of the cases described above: there could be a possibility that an initially failed item that was retried, has now a remote URL in the db but was still not unmarked in the Post body before the app stopped. Curious situation, but totally possible.
This case is specifically addressed and [commented in code here](https://github.com/wordpress-mobile/WordPress-Android/pull/6460/files#diff-ffbdd6cd0e03b7a13fcdb13666c6d8c3R1538)
Added another comment here just in case 2cb38f7

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 8 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: good call, addressed in 2cb38f7

**Comment by mzorz** [comment]
Ready for another round here @nbradbury 🙇

**Comment by nbradbury** [comment]
Code looks good, but while testing it just now I discovered that if I have a post with failed media, every time I open it in the editor it runs through `resetUploadingMediaToFailedIfNotInProgressOrQueued`, gets marked as an error [here](https://github.com/wordpress-mobile/WordPress-Android/pull/6460/files#diff-38b6aea9cfff5f3e0e4b7389b44de602R429), then gets marked as changed [here](https://github.com/wordpress-mobile/WordPress-Android/pull/6460/files#diff-38b6aea9cfff5f3e0e4b7389b44de602R435) and has it's local timestamp changed.

In other words, editing a post with failed media always marks it as changed. I don't think we want this, do we?

**Comment by nbradbury** [comment]
I think I found another issue. I removed failed media from a post then tapped UPDATE and saw the message below. The post was entirely empty so this shouldn't occur.

screenshot_1502292177

**Comment by nbradbury** [comment]
> running steps 7 and 8 successively don't ever trigger changes for me. Can you describe the situation a bit more, maybe there's something else that's different?

Sorry, I'm not able to repro it now :( 

I do see that `UploadService.markPostAsError(mPost);` is called every time I edit a post with failed media, but it's not changing the local timestamp except the first time. Apologies for the goose chase!

**Comment by nbradbury** [comment]
Looks good! :shipit:
